omegaconf/grammar/OmegaConfGrammarLexer.g4

```diff
@@ -60,9 +60,21 @@ ESC: (ESC_BACKSLASH | '\\(' | '\\)' | '\\[' | '\\]' | '\\{' | '\\}' |
       '\\:' | '\\=' | '\\,' | '\\ ' | '\\\t')+;
 WS: [ \t]+;
 
+// Quoted values for both types of quotes.
+// A quoted value is made of the enclosing quotes, and either:
+//   - nothing else
+//   - an even number of backslashes (meaning they are escaped)
+//   - a sequence of any character, followed by any non-backslash character, and optionally
+//     an even number of backslashes (i.e., also escaped)
+// Examples (right hand side: expected content of the resulting string, after processing):
+//    ""                    -> <empty>
+//    '\\'                  -> \
+//    "\\\\"                -> \\
+//    'abc\\'               -> abc\
+//    "abc\"def\\\'ghi\\\\" -> abc"def\\\'ghi\\
 QUOTED_VALUE:
-      '\'' ('\\\''|.)*? '\'' // Single quotes, can contain escaped single quote : /'
-    | '"' ('\\"'|.)*? '"' ;  // Double quotes, can contain escaped double quote : /"
+      '"' (('\\\\')* | (.)*? ~[\\] ('\\\\')*) '"'     // double quotes
+    | '\'' (('\\\\')* | (.)*? ~[\\] ('\\\\')*) '\'';  // single quotes
```

Comment:
The comments above are meant to explain it => I guess not clear enough?

Maybe it's missing an explanation of the main idea, which is that we want the string to finish by a quote optionally preceded by an even number of `\`. Would that help? Otherwise, suggestion to make it clearer?

Reply:
The comment is explaining the behavior.
I am asking about the token definition.
Usually I can read tokens/parser rules and get a good about what to expect. this is not the case here.

Reply:
(I am not asking you to change anything in the comment)

Reply:
Here's the full rule as a reminder:
```
QUOTED_VALUE:
      '"' (('\\\\')* | (.)*? ~[\\] ('\\\\')*) '"'     // double quotes
    | '\'' (('\\\\')* | (.)*? ~[\\] ('\\\\')*) '\'';  // single quotes
```

The comments above explain the different chunks in order. Let's go through them:

> Quoted values for both types of quotes.

This is why there are two lines (one for double quotes and one for single quotes).
In the following, let's look only at the first line.

> A quoted value is made of the enclosing quotes...

These are the first and last elements in the expression: `'"'` ... `'"'`

> ... and either:
> - nothing else
> - an even number of backslashes (meaning they are escaped)

These are captured by the first option for the content of the string, matched by: `('\\\\')*`

> - a sequence of any character, followed by any non-backslash character, and optionally an even number of backslashes (i.e., also escaped)

This is captured by the second option for the content of the string, matched by: `(.)*? ~[\\] ('\\\\')*`. We have respectively:
* a sequence of any character: `(.)*?`
* any non-backslash character: `~[\\]`
* optionally an even number of backslashes: `('\\\\')*`

Let me know if that helps. Also if you can think of a simpler way to express the same logic, happy to take it!

Reply:
> an even number of backslashes (meaning they are escaped)

What inputs is representing? a string with nothing but slashses?
```
\
\\
\\\
```
?

> a sequence of any character, followed by any non-backslash character, and optionally an even number of backslashes (i.e., also escaped)

It feels like this should support a string that ends with backslashes like in my example above.

Reply:
> What inputs is representing? a string with nothing but slashses?
> 
> ```
> \
> \\
> \\\
> ```
> 
> ?

A string with nothing but backslashes is captured by the first option for the content of the string: `('\\\\')*`
The number of backslashes has to be even: `"\"` isn't a valid quoted string since it doesn't have an end quote (the second quote is escaped).

> It feels like this should support a string that ends with backslashes like in my example above.

It does: either there are only backslashes and this is covered above, or there is at least one non-backslash character and this is covered by the second case. But the number of ending backslashes must always be even (for the same reason as above).